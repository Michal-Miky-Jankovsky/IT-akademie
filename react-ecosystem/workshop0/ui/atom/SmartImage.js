import observe from '@jurca/react-dom-observe';
import classnames from 'classnames';
import throttle from 'lodash.throttle';
import PropTypes from 'prop-types';
import React from 'react';
import Loader from '~/ui/atom/Loader';
import Sizer from '~/ui/atom/Sizer';
import {getBoundingClientRect, getVisibilityRatio} from '~/util/componentHelpers';

const EXTENDED_PADDING = 300;

export default class SmartImage extends React.PureComponent {
  static propTypes = {
    src: PropTypes.string.isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    layout: PropTypes.string,
    alt: PropTypes.string,
  };

  static defaultProps = {
    width: null,
    height: null,
  };

  constructor(props) {
    super(props);

    this.state = {
      noloading: false,
      visibleInViewport: false,
    };

    this._root = null;
    this._throttledCheckVisibility = throttle(() => this._checkVisibility(), 333);
  }

  render() {
    return (
      <div
        className={classnames(
          'atm-image',
          'atm-overflow',
          {
            'atm-responsive': this.props.layout === 'responsive',
          },
        )}
        style={this.props.layout !== 'responsive' ?
          {
            width: this.props.width,
            height: this.props.height,
          }
          :
          null
        }
        ref={null/* TODO: obtain the root element reference */}
      >
        {this.props.layout === 'responsive' &&
          <Sizer
            width={this.props.width}
            height={this.props.height}
            placeholder
          />
        }
        {/* TODO: show loader while loading instead of img */}
        <img
          src={this.props.src}
          alt={this.props.alt}
          className={classnames(
            'atm-fill',
            {
              'atm-loaded': this.state.noloading,
            },
          )}
        />
      </div>
    );
  }

  _rootMounted(rootNode) {
    this._root = rootNode;
    this._bindEventListeners();
    this._checkVisibility();
  }

  _rootUnmounted() {
    this._root = null;
    this._unbindEventListeners();
  }

  _bindEventListeners() {
    window.addEventListener('resize', this._throttledCheckVisibility);
    window.addEventListener('scroll', this._throttledCheckVisibility);
  }

  _unbindEventListeners() {
    window.removeEventListener('resize', this._throttledCheckVisibility);
    window.removeEventListener('scroll', this._throttledCheckVisibility);
  }

  _checkVisibility() {
    if (!this._root) {
      return;
    }

    // TODO - see the pseudocode below

    // get bounds of the root element with some extra padding

    // if the element is visible and we did not trigger image loading yet:
    //   - load image
    //   - remove event listeners
    //   - set flag that we started image loading
  }

  _preLoadImage() {
    const image = new Image();

    image.onload = () => {
      if (this._root) {
        this.setState({noloading: true});
      }
    };
    image.onerror = () => {
      if (this._root) {
        this.setState({noloading: true});
      }
    };

    if (this.props.src) {
      image.src = this.props.src;
    } else {
      if (this._root) {
        this.setState({noloading: true});
      }
    }
  }
}
