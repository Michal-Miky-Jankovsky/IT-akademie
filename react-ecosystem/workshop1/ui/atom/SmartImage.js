import observe from '@jurca/react-dom-observe';
import classnames from 'classnames';
import throttle from 'lodash.throttle';
import PropTypes from 'prop-types';
import Loader from '~/ui/atom/Loader';
import Sizer from '~/ui/atom/Sizer';
import {getBoundingClientRect, getVisibilityRatio} from '~/util/componentHelpers';

const EXTENDED_PADDING = 300;

export default class SmartImage extends React.PureComponent {
  static propTypes = {
    src: PropTypes.string.isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    layout: PropTypes.string,
    alt: PropTypes.string,
  };

  static defaultProps = {
    width: null,
    height: null,
  };

  constructor(props) {
    super(props);

    this.state = {
      noloading: false,
      visibleInViewport: false,
    };

    this._root = null;
    this._throttledCheckVisibility = throttle(() => this._checkVisibility(), 333);
  }

  render() {
    return (
      <div
        className={classnames(
          'atm-image',
          'atm-overflow',
          {
            'atm-responsive': this.props.layout === 'responsive',
          },
        )}
        style={this.props.layout !== 'responsive' ?
          {
            width: this.props.width,
            height: this.props.height,
          }
          :
          null
        }
        ref={observe(this, this._rootMounted, this._rootUnmounted)}
      >
        {this.props.layout === 'responsive' &&
          <Sizer
            width={this.props.width}
            height={this.props.height}
            placeholder
          />
        }
        {this.state.noloading ?
          <img
            src={this.props.src}
            alt={this.props.alt}
            className={classnames(
              'atm-fill',
              {
                'atm-loaded': this.state.noloading && this.state.visibleInViewport,
              },
            )}
          />
          :
          <Loader mode="small" layout="center"/>
        }
      </div>
    );
  }

  _rootMounted(rootNode) {
    this._root = rootNode;
    this._bindEventListeners();
    this._checkVisibility();
  }

  _rootUnmounted() {
    this._root = null;
    this._unbindEventListeners();
  }

  _bindEventListeners() {
    window.addEventListener('resize', this._throttledCheckVisibility);
    window.addEventListener('scroll', this._throttledCheckVisibility);
  }

  _unbindEventListeners() {
    window.removeEventListener('resize', this._throttledCheckVisibility);
    window.removeEventListener('scroll', this._throttledCheckVisibility);
  }

  _checkVisibility() {
    if (!this._root) {
      return;
    }

    const rootElementRect = getBoundingClientRect(
      this._root,
      {
        width: this.props.width,
        height: this.props.height,
      },
      EXTENDED_PADDING,
    );

    if (
      this.state.visibleInViewport === false &&
      getVisibilityRatio(rootElementRect) > 0
    ) {
      this._preLoadImage();
      this._unbindEventListeners();
      this.setState({visibleInViewport: true});
    }
  }

  _preLoadImage() {
    const image = new Image();

    image.onload = () => {
      if (this._root) {
        this.setState({noloading: true});
      }
    };
    image.onerror = () => {
      if (this._root) {
        this.setState({noloading: true});
      }
    };

    if (this.props.src) {
      image.src = this.props.src;
    } else {
      if (this._root) {
        this.setState({noloading: true});
      }
    }
  }
}
