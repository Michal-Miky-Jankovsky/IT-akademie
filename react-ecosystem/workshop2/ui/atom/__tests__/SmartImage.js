import React from 'react';
import Renderer from 'react-test-renderer';
import SmartImage from '~/ui/atom/SmartImage';

describe('SmartImage', () => {
  it('should render correctly', () => {
    const smartImage = <SmartImage src="//some/image.png" alt="image" width={320} height={200} layout="responsive"/>;
    const domTree = Renderer.create(smartImage).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
