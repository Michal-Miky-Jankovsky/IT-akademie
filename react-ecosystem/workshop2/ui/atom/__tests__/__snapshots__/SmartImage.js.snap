// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartImage should render correctly 1`] = `
<div
  className="atm-image atm-overflow atm-responsive"
  style={null}
>
  <div
    className="jsx-151239249 atm-sizer atm-placeholder"
  />
  <div
    className="atm-loader atm-loader-small atm-loader-center"
  >
    <span />
    <span />
    <span />
    <span />
    <span />
  </div>
</div>
`;
